autoprecompiles/src/equivalence_classes.rs

```diff
@@ -2,74 +2,164 @@ use std::collections::{BTreeSet, HashMap};
+        let mut class_of = HashMap::new();
+        let mut num_classes = 0;
+
+        for class in iter {
+            let class_iter = class.into_iter();
+            // Ignore classes with 0 or 1 elements as they are useless
+            if class_iter.len() > 1 {
+                for element in class_iter {
+                    class_of.insert(element, num_classes);
+                }
+                num_classes += 1;
+            }
+        }
+
         Self {
-            inner: iter.into_iter().filter(|class| class.len() > 1).collect(),
+            class_of,
+            num_classes,
         }
     }
 }
 
-impl<T> Partition<T> {
-    pub fn iter(&self) -> impl Iterator<Item = &EquivalenceClass<T>> {
-        self.inner.iter()
+impl<T: Eq + Hash + Clone> Partition<T> {
+    /// Returns all equivalence classes as a Vec<Vec<T>>.
+    /// This is O(n) where n is the number of elements.
+    #[allow(clippy::iter_over_hash_type)] // Order within classes doesn't matter semantically
+    pub fn into_classes(&self) -> Vec<Vec<T>> {
+        let mut classes: Vec<Vec<T>> = vec![Vec::new(); self.num_classes];
+        for (elem, &class_id) in &self.class_of {
+            classes[class_id].push(elem.clone());
+        }
+        classes
     }
 }
 
-impl<T: Ord + Hash + Copy> Partition<T> {
+impl<T: Eq + Hash + Copy> Partition<T> {
     /// Intersects multiple partitions of the same universe into a single partition.
     /// In other words, two elements are in the same equivalence class in the resulting partition
     /// if and only if they are in the same equivalence class in all input partitions.
     /// Singleton equivalence classes are omitted from the result.
     pub fn intersect(partitions: &[Self]) -> Self {
-        // For each partition, build a map: Id -> class_index
-        let class_ids: Vec<HashMap<T, usize>> = partitions
+        let Some(first) = partitions.first() else {
+            return Self {
+                class_of: HashMap::new(),
+                num_classes: 0,
+            };
+        };
+
+        // Pairwise intersection: fold over partitions, intersecting two at a time.
+        // This is more efficient than building Vec<usize> signatures because:
+        // 1. We only hash (usize, usize) tuples instead of Vec<usize>
+        // 2. The result shrinks after each intersection, making later steps faster
+        partitions[1..]
             .iter()
-            .map(|partition| {
-                partition
-                    .iter()
-                    .enumerate()
-                    .flat_map(|(class_idx, class)| class.iter().map(move |&id| (id, class_idx)))
-                    .collect()
-            })
-            .collect();
+            .fold(first.clone(), |acc, p| Self::intersect_two(&acc, p))
+    }
 
-        // Iterate over all elements in the universe
-        partitions
+    /// Intersects two partitions.
+    fn intersect_two(a: &Self, b: &Self) -> Self {
```

Comment:
unusual to see a method take `a: &Self` instead of `&self` but I guess it works :)